A52-xds-custom-lb-policies.md

```diff
@@ -0,0 +1,312 @@
+
+The CDS resource is rejected and a NACK returned to the control plane in either of the following cases:
+
+- xDS LB policy registry cannot convert the configuration and returns an error 
+- gRPC LB policy registry cannot parse the converted configuration.
+
+### xDS LB Policy Registry
+
+This new registry will maintain a set of converters that are able to map from the xDS LoadBalancingPolicy message to the internal gRPC JSON format. The conversion process will be recursive so that e.g. the conversion of the new xDS WRR Locality policy (see below) will call into the registry to convert the child policies.
+
+The registry will provide the following function:
+
+```javascript
+// Converts an xDS cluster load balancing policy message to the internal
+// PolicyConfig JSON string.
+//
+// An error/exception will be returned if the message type
+// is not recognized, or if policy configuration is invalid in a way that
+// can't be determined by looking at the PolicyConfig output. 
+string convertToServiceConfig(envoy.config.cluster.v3.LoadBalancingPolicy loadBalancingPolicy)
+```
+
+This function will iterate over the list of `Policy` messages in `LoadBalancingPolicy`, attempting to convert each one to gRPC form, stopping at the first supported policy. The `Policy` message contains a `TypedExtensionConfig` message with the configuration information. `TypedExtensionConfig` in turn uses an Any typed `typed_config` field to store policy configuration of any type. This `typed_config` field is used to determine both the name of a policy and the configuration for it, depending on its type:
+
+- `envoy.extensions.load_balancing_policies.ring_hash.v3.RingHash`
+  - The gRPC policy name will be `ring_hash_experimental`.
+  - The `minimum_ring_size` and `maximum_ring_size` xDS fields are converted to the `minRingSize` and `maxRingSize` fields in service config
+  - The `hash_function` field will not be converted, but will need to be validated. Any function besides `XX_HASH` will cause the conversion to fail
+- `envoy.extensions.load_balancing_policies.round_robin.v3.RoundRobin`
+  - The gRPC policy name will be `round_robin`.
+  - Has no configuration - an empty JSON object is returned
+- `envoy.extensions.load_balancing_policies.round_robin.v3.WrrLocality`
+  - The gRPC policy name will be `xds_wrr_locality_experimental`
+  - Contains a `LoadBalancingPolicy` field that has a list of endpoint picking policies. This field will be processed by recursively calling the xDS LB policy registry, and the result will be used in the `child_policy` field in the configuration for the `xds_wrr_locality_experimental` policy (see the “xDS WRR Locality Load Balancer” sections for details on the config format).
+- `xds.type.v3.TypedStruct`
+  - This type represents a custom policy, deployed with the gRPC client by a user
+  - The gRPC policy name will be value of the `type_url` field in the `TypedStruct`, with the "`type.googleapis.com/`" prefix stripped out
```

Comment:
As we've discussed, the prefix here is weird. We probably want to follow protobuf here and use everything after the last `/`.

https://github.com/protocolbuffers/protobuf/blob/5f920d15071a160c65745568a85315cb6fd61600/src/google/protobuf/compiler/java/message.cc#L1687 (powers the [`is(Class)` function](https://github.com/protocolbuffers/protobuf/blob/5f920d15071a160c65745568a85315cb6fd61600/src/google/protobuf/compiler/java/message.cc#L1717))

Reply:
Updated.